witness/src/Dscp/Snowdrop/Validators.hs

```diff
@@ -28,22 +30,55 @@ validator = mempty
 instance
     SD.VerifySign
         PublicKey
-        (Signature (TxId, PublicKey))
-        (TxId, PublicKey)
+        (Signature (TxId, PublicKey, ()))
+        (TxId, PublicKey, ())
+  where
+    verifySignature = verify
+
+type ThePublication = Publication (Hash PrivateBlock) Address
+
+instance
+    SD.VerifySign
+        PublicKey
+        (Signature (TxId, PublicKey, ThePublication))
+        (TxId, PublicKey, ThePublication)
   where
     verifySignature = verify
 
 instance SD.HasGetter PublicKey Address where
     gett = Address . hash
 
-instance SD.HasGetter Proofs (SD.WithSignature PublicKey (Signature (TxId, PublicKey)) (TxId, PublicKey)) where
-    gett (AddressTxWitness it) = it
+instance SD.HasPrism Proofs AddrTxProof where
+    proj (AddressTxWitness it) = Just it
+    proj  _                    = Nothing
+
+instance SD.HasReview Proofs AddrTxProof where
+    inj = AddressTxWitness
+
+instance SD.HasPrism Proofs PublicationTxProof where
+    proj (PublicationTxWitness it) = Just it
+    proj  _                        = Nothing
+
+instance SD.HasReview Proofs PublicationTxProof where
+    inj = PublicationTxWitness
+
+instance SD.HasPrism Proofs TxId where
+    proj (AddressTxWitness (SD.WithSignature { wsBody = (it, _, _) })) = Just it
+    proj  _                                                            = Nothing
+
+instance SD.HasReview Proofs TxId where
+    inj = error "impossible to implement"
```

Comment:
This doesn't seem right. What's going on?

Reply:
Because `class HasReview a b => HasPrism a b`. We can't have `proj` w/o `inj`, thanks to SD.

And we can't have `HasGetter`, because there is a case where we have no `TxId` in `Proofs`.

This should be solved on SD side, by making `class (HasView a b, HasReview a b) => HasPrism a b`.

Reply:
@Heimdell I suggest you fix it on SD side -- I will review & merge SD PR immediately, so it won't block this PR.

Reply:
So you won't? Ok, at least specify the reason in the code comment.